### PR TITLE
Modify Contribution file

### LINKED ISSUE #12: Add PR review process and general acceptance criteria to contributing guide
The contributing guide needs to define the process for reviewing a PR and the quality of the commits.

Some suggestions:
- Descriptive commit message titles and bodies, explaining what the PR is about and if it depends on any other unmerged PRs
- Students-driven quality review cycles (no LGTMs) and only then mentors provide their reviews.
- Code review needs to be according to the task requirements - i.e. you need to know the acceptance criteria of the task before reviewing.
- Code needs to be merged before the weekend.

### PR DESCRIPTION
**Pull request by pazMenachem**
Review requirements added.
Commit guide lines added. (under "commit your update")
Pull request guide lines modified.
fix #12

### CODE REVIEW

**Comment by RawanAbuleil** [comment]
looks good , great work paz .

**Comment by thetaxas2468** [comment]
LGTM!

**Comment by MeitarRizner** [comment]
Good job!

**Comment by yftacherzog** [comment]
Got to appreciate the irony that a PR defining that reviews cannot be all "lgtm" have three lgtm reviews.

**Comment by yftacherzog** [comment]
@RawanAbuleil @thetaxas2468 @MeitarRizner @LiorSilberman 
We're going to need some more compelling reviews before the mentors can provide their reviews.

**Review by LiorSilberman — COMMENTED** [review]
You can delete some extra empty lines that you have in the file.

in pull request:
maybe we can add that after uploading PR, you should let the team members know that there are new PR by writing in slack or by direct message if the PR directed for someone.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
+try to be descriptive as you can while keeping it simple.
+
+- Example:
+Adding an awesome feature.
+
+Adding feature which:
+1. makes cookies.
+2. makes you get A+ in all courses.
+3. create dank memes.
+in order to use each of the feature above just go to page ****,
```
Comment (thetaxas2468) [inline CONTRIBUTING.md:36]: No need for example here i think since all details has been written

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
 ---
+## Review requirements 
+- Before revewing the task, please acknowledge its requirements and acceptance criteria.
+
 
+--- 
 ## Pull Request
 - When you're finished with the changes, create a pull request, also known as a PR.
-- Please note that at least two team members of the group should review and approve the pull request before asking the mentors to review and merge.
+- Please note that at least two team members of the group should quality review(NO LGTM) and approve the pull request before asking the mentors to review and merge.
```
Comment (thetaxas2468) [inline CONTRIBUTING.md:50]: NO LGTM ,should be betwen us but not in public i think you should delete it

File: CONTRIBUTING.md (hunk 1 above)
Comment (nofaralfasi) [reply CONTRIBUTING.md:36]: Please notice that you commented on lines 30-36 (I think you meant to mark lines 29-30).

**Comment by nofaralfasi** [comment]
> You can delete some extra empty lines that you have in the file.
>
> in pull request: maybe we can add that after uploading PR, you should let the team members know that there are new PR by writing in slack or by direct message if the PR directed for someone.

It is better to mark the lines you are referring to in the code and write your comments there (as @thetaxas2468 did in this PR).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
+- Example:
+Adding an awesome feature.
+
+Adding feature which:
+1. makes cookies.
+2. makes you get A+ in all courses.
+3. create dank memes.
+in order to use each of the feature above just go to page ****,
+write the line "make_cookies()" where ever you want and walla, 
+cookies will appear.
```
Comment (nofaralfasi) [inline CONTRIBUTING.md:38]: For me, this example wasn't clear. 
First of all, I would write it inside a code block: `Example: this is a code block.`
And it is a little bit too long for me. The purpose here is to show how to write a good commit message.

**Comment by nofaralfasi** [comment]
Hi @pazMenachem, here are a few notes:

1. I suggest adding more to the `Making changes locally` section. Maybe even renaming it `Prerequisites`. 
2. Some of the titles can be better described (as I mentioned above). 
I suggest looking at older projects from previous courses for more examples.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
+
 Commit the changes once you are happy with them. 
 
 ---
+## Review requirements 
+- Before revewing the task, please acknowledge its requirements and acceptance criteria.
+
```
Comment (yftacherzog) [inline CONTRIBUTING.md:45]:
```suggestion
- When reviewing, confirm that the code accomplishes the requirements and the acceptance criteria, and that the accompanying unit tests cover all the relevant edge cases.

```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
 
+--- 
 ## Pull Request
 - When you're finished with the changes, create a pull request, also known as a PR.
-- Please note that at least two team members of the group should review and approve the pull request before asking the mentors to review and merge.
+- Please note that at least two team members of the group should quality review(NO LGTM) and approve the pull request before asking the mentors to review and merge.
 - The reviewer resolve the conversation to indicate that they are ok with the updated content.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:51]:
```suggestion
- The reviewer resolves the conversation to indicate that they are ok with the updated content.
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (Omeramsc) [reply CONTRIBUTING.md:50]: Nothing should be kept private between you. You should consider this project as any open source project; if someone new would like to join the effort - this doc should let him know everything about how to contribute. CR process included.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -18,13 +18,36 @@ Welcome! We're so excited that you're considering contributing to the official G
...
 - If you run into any merge issues, check how to solve them and resolve merge conflicts and other issues.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:52]: Stuff which I find missing:

```
- If your PR depends on any other PRs being merged, indicate so on the commit message or on the comments. Specifically, if it contains unmerged commits of any other PRs, indicate to the reviewers which commits are expected to be reviewed as part of your PR.
- Keep your PRs as small as possible.
- Ideally, a single PR should include a single commit.
- A PR's commit, needs to represent a transition in the code that completes a task (e.g. a new function, along with its unit tests) .
- Changes made as part of the review process (i.e. fixes to the content of the original commit) , should be added to the original commit (e.g. using `commit --amend`), rather than creating a new commit. 
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,54 +1,57 @@
...
+- Set up a development environment to begin working on the project, all you need to do is 'vagrant up'
 - Start with your changes!
-
 ---
-
-## Commit your update
-Commit the changes once you are happy with them. 
+## Commiting
+Please use the guide below inorder to follow our commit message convention:
+* [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
```
Comment (yftacherzog) [inline CONTRIBUTING.md:13]: This very commit does not follow the guide that you reference. e.g. the commit message title does not follow that format.

Also, I don't want the body to be optional. It should be mandatory.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,54 +1,57 @@
...
+- If your PR depends on the merging of other PRs, indicate this on the commit message or on the comments.
+Specifically, if it contains unmerged commits of any other PRs,
+indicate to the reviewers which commits are expected to be reviewed as part of your PR.
+- Keep your PRs as small as possible.
+Ideally, a single PR should include a single commit.
+- The reviewer resolves the conversation to indicate that they are ok with the updated content.
 - If you run into any merge issues, check how to solve them and resolve merge conflicts and other issues.
-
+- Changes made as part of the review process (i.e. fixes to the content of the original commit) ,
+should be added to the original commit (e.g. using ```commit --amend```), rather than creating a new commit.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:26]:
```suggestion
should be added to the original commit (e.g. using `commit --amend`), rather than creating a new commit. 
```
You should use this triple thingy when you're adding a code block. When adding inline code, you just do a single-thingy.

```
the block
is hot
```

vs. `the fox is black`.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,54 +1,57 @@
...
+indicate to the reviewers which commits are expected to be reviewed as part of your PR.
+- Keep your PRs as small as possible.
+Ideally, a single PR should include a single commit.
+- The reviewer resolves the conversation to indicate that they are ok with the updated content.
 - If you run into any merge issues, check how to solve them and resolve merge conflicts and other issues.
-
+- Changes made as part of the review process (i.e. fixes to the content of the original commit) ,
+should be added to the original commit (e.g. using ```commit --amend```), rather than creating a new commit. 
+- Please keep in mind that at least two team members of the group should "quality review"(i.e. NO LGTM) and approve the pull request before asking the mentors to review and merge it.
+- Each PR should include testing(Code testing section below).
```
Comment (yftacherzog) [inline CONTRIBUTING.md:28]:
```suggestion
- Each PR should include testing (Code testing section below).
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,54 +1,57 @@
...
 ---
+## Code Testing
+Please use Pytest as your test framework.
+
+* [Pytest](https://docs.pytest.org/en/7.2.x/getting-started.html)
+- Please note that Pytest is already installed in your virtual environment.
 
+---
+## Review requirements 
+- When reviewing, confirm that the code accomplishes the requirements and the acceptance criteria,
```
Comment (yftacherzog) [inline CONTRIBUTING.md:39]: Need to define, the requirements and acceptance criteria of what. GitHub issue? commit message? Something else?

If it's a GitHub issue, then you should define that each PR should be [tied to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue). If it's the commit message, then you need more elaborated criteria for the commit message content.

I think linking PRs to issues is easier to accomplish and more flexible.

File: CONTRIBUTING.md (hunk 7 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:13]: you're right, i didnt notice those details.

File: CONTRIBUTING.md (hunk 8 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:26]: commit --amend doesnt count as code block?

File: CONTRIBUTING.md (hunk 10 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:39]: i think i will just delete this section, i dont find it much informative and i dont know how to elaborate it properly.

File: CONTRIBUTING.md (hunk 8 above)
Comment (yftacherzog) [reply CONTRIBUTING.md:26]: If it's on a line of its own (or multiple lines), you use \`\`\`. If it's together on the same line with regular text you use a single \`

File: CONTRIBUTING.md (hunk 10 above)
Comment (yftacherzog) [reply CONTRIBUTING.md:39]: but that is how you decide whether the PR is doing what it's suppose to be doing or not. If you don't have that, then you can't tell whether the PR is actually doing what it's supposed to be doing.

Comment (yftacherzog) [reply CONTRIBUTING.md:39]: You can say here that the PR needs to either be linked to a GitHub issue, or should specify in details what it's supposed to be achieving.

Comment (pazMenachem) [reply CONTRIBUTING.md]: added your notes, thanks!

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,54 +1,59 @@
...
 - Start with your changes!
-
 ---
-
-## Commit your update
-Commit the changes once you are happy with them. 
-
+## Commiting
+Please use the guide below inorder to follow our commit message convention:
+* [commit convention](https://cbea.ms/git-commit/#seven-rules)
```
Comment (yftacherzog) [inline CONTRIBUTING.md:13]: I think this standard is a good fit. Just note that the commit message for this MR doesn't comply with it either. Specifically, [this point](https://cbea.ms/git-commit/#imperative). The commit message should be something like: `Modify contributing file`.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,54 +1,59 @@
...
-- The reviewer resolve the conversation to indicate that they are ok with the updated content.
+- If your PR depends on the merging of other PRs, indicate this on the commit message or on the comments.
+Specifically, if it contains unmerged commits of any other PRs,
+indicate to the reviewers which commits are expected to be reviewed as part of your PR.
+- Keep your PRs as small as possible.
+Ideally, a single PR should include a single commit.
+- The reviewer resolves the conversation to indicate that they are ok with the updated content.
 - If you run into any merge issues, check how to solve them and resolve merge conflicts and other issues.
-
+- Changes made as part of the review process (i.e. fixes to the content of the original commit) ,
```
Comment (yftacherzog) [inline CONTRIBUTING.md:24]:
```suggestion
- Changes made as part of the review process (i.e. fixes to the content of the original commit),
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,54 +1,59 @@
...
 ---
+## Code Testing
+Please use Pytest as your test framework.
+
+* [Pytest](https://docs.pytest.org/en/7.2.x/getting-started.html)
+- Please note that Pytest is already installed in your virtual environment.
 
+---
+## Review Requirements
+When reviewing, confirm that the code accomplishes the requirements and the acceptance criteria(Flake8, commit convention, etc.).
```
Comment (yftacherzog) [inline CONTRIBUTING.md:39]:
```suggestion
When reviewing, confirm that the code accomplishes the requirements and the acceptance criteria (Flake8,
commit convention, etc.).
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,54 +1,59 @@
...
+A GitHub issue: the code fixes the problem.
+Adding a feature: elaborate in the commit the purpose of the code.
+Both cases should have accompanying unit tests that cover all the relevant edge cases.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:42]: Something is broken with the format of this paragraph. You can preview the file on GitHub or on an editor (online or on your machine), to see it.

File: CONTRIBUTING.md (hunk 13 above)
Comment (yftacherzog) [reply CONTRIBUTING.md:39]: also, flake8 and commit conventions are not the acceptance criteria. The acceptance criteria is what this PR should accomplish (should be described on an issue or on the commit message - or both).

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,54 +1,59 @@
...
-- Fork the repo so that you can make your changes without affecting the original project until you're ready to merge them.
-
-- Create a working branch 
-
-- Set up a development environment to begin working on the project all you need to do is 'vagrant up'
-
+## Prerequisites
+- Fork the repo to make your changes without affecting the original project until you're ready to merge.
+- Create a working branch.
+- Set up a development environment to begin working on the project, all you need to do is 'vagrant up'
```
Comment (yftacherzog) [inline CONTRIBUTING.md:8]:
```suggestion
- Set up a development environment to begin working on the project, all you need to do is 'vagrant up'.
```

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,54 +1,59 @@
...
 - Start with your changes!
-
 ---
-
-## Commit your update
-Commit the changes once you are happy with them. 
-
+## Commiting
+Please use the guide below inorder to follow our commit message convention:
```
Comment (yftacherzog) [inline CONTRIBUTING.md:12]:
```suggestion
Please use the guide below in order to follow our commit message convention:
```

File: CONTRIBUTING.md (hunk 11 above)
Comment (yftacherzog) [reply CONTRIBUTING.md:13]: (to be clear, I think you should change this commit message rather than the standard)

**Comment by yftacherzog** [comment]
@pazMenachem almost there. Good job!

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -1,54 +1,59 @@
...
-## Make changes locally
-
-- Fork the repo so that you can make your changes without affecting the original project until you're ready to merge them.
-
-- Create a working branch 
-
-- Set up a development environment to begin working on the project all you need to do is 'vagrant up'
-
+## Prerequisites
+- Fork the repo to make your changes without affecting the original project until you're ready to merge.
```
Comment (Omeramsc) [inline CONTRIBUTING.md:6]: ready to propose your code changes*

The maintainers are the ones who should be ready to merge

File: CONTRIBUTING.md (hunk 15 above)
Comment (Omeramsc) [inline CONTRIBUTING.md:8]: Hey, I'm a new developer who wish to contribute to this project! I ran `vagrant up` like you suggested, and got `bash: vagrant: command not found`, any idea what the problem it? :)

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -1,54 +1,59 @@
...
+## Commiting
+Please use the guide below inorder to follow our commit message convention:
+* [commit convention](https://cbea.ms/git-commit/#seven-rules)
 ---
-
 ## Pull Request
 - When you're finished with the changes, create a pull request, also known as a PR.
-- Please note that at least two team members of the group should review and approve the pull request before asking the mentors to review and merge.
-- The reviewer resolve the conversation to indicate that they are ok with the updated content.
+- If your PR depends on the merging of other PRs, indicate this on the commit message or on the comments.
```
Comment (Omeramsc) [inline CONTRIBUTING.md:17]: don't indicate it in the comment section, indicate it on the commit message. 
also indicate it as such: `depends on #<number of PR>` so it'll be linked.

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -1,54 +1,59 @@
...
+Specifically, if it contains unmerged commits of any other PRs,
+indicate to the reviewers which commits are expected to be reviewed as part of your PR.
+- Keep your PRs as small as possible.
+Ideally, a single PR should include a single commit.
+- The reviewer resolves the conversation to indicate that they are ok with the updated content.
 - If you run into any merge issues, check how to solve them and resolve merge conflicts and other issues.
-
+- Changes made as part of the review process (i.e. fixes to the content of the original commit) ,
+should be added to the original commit (e.g. using `commit --amend`), rather than creating a new commit. 
+- Please keep in mind that at least two team members of the group should "quality review" (i.e. NO LGTM)
```
Comment (Omeramsc) [inline CONTRIBUTING.md:26]: `should provide a sufficient feedback` seems better IMO. there is a time and a place for LGTM, but it can't be the case in most of the time.

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -1,54 +1,59 @@
...
+and approve the pull request before asking the mentors to review and merge it.
```
Comment (Omeramsc) [inline CONTRIBUTING.md:27]: s/mentors/maintainers.

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -1,54 +1,59 @@
...
+- Each PR should include testing (Code testing section below).
```
Comment (Omeramsc) [inline CONTRIBUTING.md:28]: should include tests for the features it presents*

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -1,54 +1,59 @@
...
+PR without tests won't be reviewed.
 ---
+## Code Testing
+Please use Pytest as your test framework.
```
Comment (Omeramsc) [inline CONTRIBUTING.md:32]: As you probably won't accept changes with different test frameworks, you don't have to be polite. `This project uses Pytest as its test framework`

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -1,54 +1,59 @@
...
+
+* [Pytest](https://docs.pytest.org/en/7.2.x/getting-started.html)
+- Please note that Pytest is already installed in your virtual environment.
```
Comment (Omeramsc) [inline CONTRIBUTING.md:35]: In this project's virtual environment*

**Comment by Omeramsc** [comment]
- I would suggest including information about the development environment in the README file, as it is mentioned multiple times (what is vagrant? for example). this is for another PR, but make sure to at least open an issue for it.
- Do mention that those rules are enforced. PRs which doesn't follow those rules won't be reviewed (this PR does not follow the guide... )

**Comment by Omeramsc** [comment]
also:
- When creating a PR, assign yourself to it and assign reviewers so they'll be notified.
- mention the issue the PR is solving, like this: ' fix #12 ' (this PR lacks that), it'll link the Issue and the PR, so that when the PR merges, the issue will be automatically closed.

File: CONTRIBUTING.md (hunk 12 above)
Comment (nofaralfasi) [inline CONTRIBUTING.md:24]:
```suggestion
- Changes made as part of the review process (i.e. fixes to the content of the original commit),
```

File: CONTRIBUTING.md (hunk 11 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:13]: will change the title

File: CONTRIBUTING.md (hunk 15 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:8]: installing vagrant, right...✔

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -1,54 +1,59 @@
...
+- Create a working branch.
+- Set up a development environment to begin working on the project, all you need to do is 'vagrant up'
 - Start with your changes!
-
 ---
-
-## Commit your update
-Commit the changes once you are happy with them. 
-
+## Commiting
```
[the remark this answers is not in the document]
Comment (pazMenachem) [reply CONTRIBUTING.md:11]: the section still exists, it is small, with a reference to outer guide, but still...
i think it answers on how to make a commit.

File: CONTRIBUTING.md (hunk 20 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:27]: care to explain? s?

File: CONTRIBUTING.md (hunk 22 above)
Comment (pazMenachem) [reply CONTRIBUTING.md:32]: half British🤷‍♂️

**Comment by pazMenachem** [comment]
> * I would suggest including information about the development environment in the README file, as it is mentioned multiple times (what is vagrant? for example). this is for another PR, but make sure to at least open an issue for it.
> * Do mention that those rules are enforced. PRs which doesn't follow those rules won't be reviewed (this PR does not follow the guide... )

ill open an issue for it.

care to mention where else it doesnt follow the guide? except the title that will be changed next commit. 


> also:
> 
> * When creating a PR, assign yourself to it and assign reviewers so they'll be notified.
> * mention the issue the PR is solving, like this: ' fix [Add PR review process and general acceptance criteria to contributing guide #12](https://github.com/redhat-beyond/beyond08-group3/issues/12) ' (this PR lacks that), it'll link the Issue and the PR, so that when the PR merges, the issue will be automatically closed.

fix #12.

tried looking how to write the same you did with out success, would like to know please.

File: CONTRIBUTING.md (hunk 20 above)
Comment (yftacherzog) [reply CONTRIBUTING.md:34]: change `mentors` to `maintainers`

Comment (yftacherzog) [reply CONTRIBUTING.md:34]: https://en.wikipedia.org/wiki/Sed#Substitution_command

**Comment by Omeramsc** [comment]
FYI you don't have to create a new commit to change those details:
- PR title and description can just be edited through GH. 
- Commit details can be changed by doing `git commit --amend` -> <make changes> -> `git push -f origin <branch_name>`

**Comment by Omeramsc** [comment]
- You did the `fix` tag right, but it should be in the description of the PR, not in the comments. note that GH did not link the issue to the PR as a result.
- I'll see that you'll have right to assign each other.

**Comment by pazMenachem** [comment]
> * You did the `fix` tag right, but it should be in the description of the PR, not in the comments. note that GH did not link the issue to the PR as a result.
> * I'll see that you'll have right to assign each other.


> FYI you don't have to create a new commit to change those details:
> 
> * PR title and description can just be edited through GH.
> * Commit details can be changed by doing `git commit --amend` -> `<make changes>` -> `git push -f origin <branch_name>`

got it

**Review by yftacherzog — COMMENTED** [review]
@pazMenachem Once the points raised by @Omeramsc are addressed, I think this PR can be merged.

**Comment by pazMenachem** [comment]
can you refer to which points I'm missing?

**Comment by yftacherzog** [comment]
@pazMenachem i meant, you had some discussions going on with @Omeramsc. Once he confirms that he’s good with the result, the pr can be merged as far as I’m concerned.

**Review by Omeramsc — APPROVED** [review]
LGTM
do note you can now set yourself as assignee and set others as reviewers.
